### ISSUE TITLE
Features for 7.1

### ISSUE THREAD

**Issue by peastman** [pandegroup/openmm#1457]
This seems like a good time to discuss the feature list for 7.1.  The last time we discussed it, the plan was to make this a fairly small release focusing on the new force fields.  It would include those and anything else that happened to be ready in time, but we wouldn't hold it up for anything else.

Is everyone still OK with that plan?  Are there any other features we specifically want to include in 7.1?

**Comment by jchodera** [comment]
I know I say this literally every time, but a threaded issue with a hundred posts isn't a great way to discuss this. Things inevitably get forgotten or missed. Can we---just for 7.1---try using the "Milestone" tags/labels to tag issues for 7.1 consideration?

**Comment by peastman** [comment]
I don't understand what you mean.  We first need to discuss this and decide what features to include.  Once we decide that a particular feature is a required one for 7.1 we can tag it as such.  But tags are not in any way a substitute for a discussion.

Or to put it differently: which are the particular features that you think should be tagged for 7.1?

**Comment by jchodera** [comment]
How about this:
* Pass through the whole issue tracker and make sure everything that is a feature request, enhancement, or bugfix is correctly labeled.
* Then, it's easy for us to quickly sort to pull out these categories and nominate things to include in 7.1 in this issue tracker
* Then you can label 7.1 milestone features/bugfixes

Could that work?

**Comment by leeping** [comment]
I'd like to chime in with a feature request for enabling periodic bonded forces as discussed in #1411.  :)

**Comment by jlmaccal** [comment]
I can probably contribute the MARTINI implementation for 7.1.

The implementation reads in GROMACS top files and produces a system with the appropriate potential (switch functions, distance dependent dielectric, etc).

The last issues to address are:
- Testing
  - All energies currently match gromacs over a range of test systems including proteins and lipids
  - All forces currently match for lipid simulations
  - The forces don't match for proteins. I believe this is due to different force distributions from v-sites. I need to adapt the tests to compute the net forces and net torques on rigid groups, which should match.
- Documentation
  - Currently there is none
- Pre-processor support
  - Gromacs supports arbitrary pre-processor expressions in defines, whereas OpenMM does not. These are sometimes used by some of the automated tools for building MARTINI systems, to say e.g. this restraint force is 4x as strong as some other one.
  - For now, I think the best solution is to require users to manually edit the files to put in constants. And of course, document this requirement.

**Comment by peastman** [comment]
I haven't gotten through the whole tracker yet, but the labels should be reasonably complete now.

**Comment by jchodera** [comment]
> I haven't gotten through the whole tracker yet, but the labels should be reasonably complete now.

Thanks!  I'm going through the list now...

**Comment by jchodera** [comment]
Here's my list:

### Forcefield improvements and small molecule support
* updated AMBER forcefields with [better agreement with Amber](https://github.com/pandegroup/openmm/issues/1347)
* [support for multisite ions](https://github.com/pandegroup/openmm/issues/1032)
* updated [CHARMM forcefields](https://github.com/pandegroup/openmm/issues/1366) (especially [CHARMM36](https://github.com/pandegroup/openmm/issues/685)) which requires [support for patches](https://github.com/pandegroup/openmm/issues/1335)
* validated [residue template generator for GAFF](https://github.com/choderalab/openmoltools/blob/master/openmoltools/forcefield_generators.py#L457-L500), which may be facilitated by [adding optional bond order to `Topology`](https://github.com/pandegroup/openmm/issues/1443)
* GBSA for small molecules, which requires [better mix-and-match handling](https://github.com/pandegroup/openmm/issues/1298) and may also involve [improvements to GB warnings](https://github.com/pandegroup/openmm/issues/1374)
* [cleanup of `examples/` directory](https://github.com/pandegroup/openmm/issues/730)

### Minor interface cleanup
* [`segment_id` support](https://github.com/pandegroup/openmm/issues/1459) and whatever else is needed to bring `Topology` object more inline with `MDTraj.Topology`. cc: https://github.com/pandegroup/openmm/issues/721 and https://github.com/mdtraj/mdtraj/pull/1002
* [minimization tolerance for `Modeller.addHydrogens`](https://github.com/pandegroup/openmm/issues/1456)
* [additional harmonized platform properties](https://github.com/pandegroup/openmm/issues/1450) and [platform precision](https://github.com/pandegroup/openmm/issues/909)
* [temperature/pressure setting interface cleanup](https://github.com/pandegroup/openmm/issues/1440); cc: https://github.com/pandegroup/openmm/issues/324
* [kinetic energy retrieval for `CustomIntegrator`](https://github.com/pandegroup/openmm/issues/1259)
* [generalization of `Modeller.addSolvent` to allow other solvents](https://github.com/pandegroup/openmm/issues/1280)
* [warning if residues have no bonds](https://github.com/pandegroup/openmm/issues/1170)
* [documented maximum random seed size](https://github.com/pandegroup/openmm/issues/1013)
* [updated documentation for `DrudeLangevinIntegrator`](https://github.com/pandegroup/openmm/issues/915)
* [retrieval of random number seeds](https://github.com/pandegroup/openmm/issues/767)
* [`Context` constructor could support `platform=None` argument](https://github.com/pandegroup/openmm/issues/639)
* [update of `CMAKE_OSX_DEPLOYMENT_TARGET` to 10.9](https://github.com/pandegroup/openmm/issues/587)
* [update CMake version requirements](https://github.com/pandegroup/openmm/issues/508)
* [harmonize cutoff get/set function names for AMOEBA](https://github.com/pandegroup/openmm/issues/435)

### Other things that would be really great
* lots of people have asked for a [faster](https://github.com/pandegroup/openmm/issues/1321) [minimizer](https://github.com/pandegroup/openmm/issues/1237)
* experimental support for [python wheels](https://github.com/pandegroup/openmm/issues/1347)
* [improvements to residue template match failures](https://github.com/pandegroup/openmm/issues/1195)
* [revisiting constraint tolerance defaults](https://github.com/pandegroup/openmm/issues/927)

**Comment by swails** [comment]
> Pre-processor support
> Gromacs supports arbitrary pre-processor expressions in defines, whereas OpenMM does not. These are sometimes used by some of the automated tools for building MARTINI systems, to say e.g. this restraint force is 4x as strong as some other one.
> For now, I think the best solution is to require users to manually edit the files to put in constants. And of course, document this requirement.

@jlmaccal -- ParmEd may have some features of interest to you here.  Check out the GROMACS support there -- it sports a dedicated C-preprocessor streaming class (`parmed/gromacs/_cpp.py`) that basically acts as a layer between the raw files and the parsers.  It may be more feature-complete than what's in OpenMM right now, and if there's stuff that it doesn't support I'd be interested in improving it (or, of course, accepting improvements :).

**Comment by jlmaccal** [comment]
@swails Cool! I'll try to take a look when I have time. I think the place to add it is to the `GromacsTopFile`, which is subclassed for Martini.

**Comment by swails** [comment]
I use it in ParmEd's `GromacsTopologyFile` class currently.  The cpp in GromacsTopFile is integrated pretty tightly into the main parser inside OpenMM, and I wanted the cpp to sit outside of that.  That's why I didn't use the OpenMM code for my own parser.  I'm certainly fine if we put _cpp.py inside OpenMM and use that in `GromacsTopFile`, but it will probably be a little bit of work to do that.

Although we could always leave the cpp stuff in place and just have that code never be triggered, I suppose...

**Comment by jchodera** [comment]
We now have a nice list of all the proposed tasks for OpenMM 7.1.   How about going through and making sure everything is labeled `enhancement`, `bug`, or `documentation`, and then selecting what you think is doable for an OpenMM 7.1 milestone or what should be left for OpenMM 7.2?  You can also feel free to assign people tasks (under Assignee), such as forcefield things we might be responsible for on our side.  This will help us all keep track of what needs to be done before the release!

**Comment by peastman** [comment]
It's not a question of what is doable.  It's a question of how big a release we want this to be.  There's no reason I *couldn't* do everything on the list.  But if I do, this will be a bigger release and take longer than what we originally planned on.  Also, @VijayPande has additional features he's interested in.  He currently has me working on Gay-Berne ellipsoids, and he's asked about DPD a few times.

So the question is, what features do we want to specifically schedule for inclusion in 7.1?  Other features might happen to get done, and if they're ready in time we'll include them.  But which are the features that have to be done before we declare 7.1 finished?

**Comment by jchodera** [comment]
What's the best way to prioritize? Maybe you could tag things with 7.1,
7.2, and 7.3 milestones after the meeting today to indicate planned
priority?

I think we had previously discussed 7.1 mostly being forcefields and small
molecule support. This division still works for me if you think focusing on
this can get out a 7.1 RC in a few weeks.

**Comment by peastman** [comment]
Perhaps we can go through the force field related things and see where we are on each of them?  Here's what I know about them:

- I've implemented part of the support for multisite ions.  It's still waiting on @rafwiewiora for other parts though.
- The implementation of patches is complete, but I'm waiting on @chayast to come up with test cases for it.
- I believe overloading is now fully implemented?

They other items I don't know about.

**Comment by jchodera** [comment]
We have to do some work on GB support for small molecules as well. I can help with that.

@rafwiewiora and @ChayaSt: Would be great if you could check in with what needs to be done or if you need more information to move forward here.

We also need to decide whether we want to include any residue template generators (eg for GAFF) with OpenMM, or if we should continue to have those in openmoltools (which is sort of an unorganized dumping ground for things at the moment).

I believe we also need to have a discussion about how the conversion scripts and source files will be merged into the OpenMM repo and how we will conduct automated testing to make sure they work. An alternative would be to have all of those conversion tools in a separate repo. But keeping scripts that are not documented or tested in the OpenMM repo isn't something we should be doing.

**Comment by rafwiewiora** [comment]
Ok, I am ready to start a first PR for the new AMBER forcefields, whenever you're ready - let me open a new issue in a sec, we should discuss things like where to put which files and how to keep backwards compatibility.

ForceField has all needed changes already, but we'd need @swails to do a ParmEd release for the amber2openmm.py to work with the public releases. 

On the TODO:
* impropers, decide on final code - tbd by tonight
* `addHydrogens` automatic search of coordinating residues
* some XML reading and `ForceField` changes to allow inputting in the info necessary for finding coordinating residues for the ions
* convert & test Zn, Ca, Mg

Let's give this 2 weeks to be complete, ok?

**Comment by ChayaSt** [comment]
I apologize for being slow about the patch tests and finishing the L-J generator. I got caught up with other projects.

Here's what is on the TODO list for CharMM:

- tests for patches (generate the systems)
- write out the patches and residues to xml files with the agreed upon design. This will be added in ParmEd but will take some time because the documentation in the CharMM files are not consistent. 
- tests for prot, carb, lipid and cgenff force fields. 

These tests will probably uncover issues so I'm not sure what the timeline for finishing this will be.

**Comment by rafwiewiora** [comment]
Based on the experience from converting the AMBER forcefield - issues keep cropping up etc., @ChayaSt most likely needs more than a few weeks to get this all right, especially with other projects going on. 

I think we should split AMBER and CHARMM for the releases - CHARMM can join at 7.2?

**Comment by peastman** [comment]
Let's see how long it takes.  "A few weeks" may be an overly aggressive schedule, and I don't think it's *that* urgent.  If taking an extra month means we can include CHARMM, I'm fine with that.

**Comment by peastman** [comment]
@rafwiewiora and @ChayaSt could you give a status update?  How are you coming on these features?

**Comment by rafwiewiora** [comment]
@peastman : I am tied to writing my PhD candidacy exam proposal by Tuesday - then I am finishing the 7.1 stuff from Wednesday onwards. Hard to predict ETA because we have the new development that is Amber Tools 16: as I am right now I need a week of work + whatever trouble the new AmberTools causes. 

14 days from now I think is a good estimate to have the AMBER part proposed, discussed and PR'ed fully.

**Comment by peastman** [comment]
Thanks!  I just wanted to check in and see what was happening.

**Comment by jchodera** [comment]
Sorry about this. Had tried to warn that both students would need to be
pulled off for critical things if we didn't finish in time. I have a
feeling much of the remainder will fall to you and me @peastman.

**Comment by peastman** [comment]
If there are specific tasks you'll need me to take over, let me know what they are.

**Comment by ChayaSt** [comment]
@peastman: as @jchodera has mentioned, I'm caught up with other projects. What I can do is finish the L-J generator since that is almost complete. I'll discuss with @jchodera what to do about the other tasks.

**Comment by ChayaSt** [comment]
Here's my plan:

* Tomorrow: clean up the L-J generator (leave some comments on the PR to point to what is complete and what is not) and hand it over to @peastman to complete. 

* Early next week: 
  1) Generate the following test systems for patches:
    * C and N termin
    * disulfide bonds
    * PMT's (dyes, phosphorylation) 
    I won't have time to write the tests so I will hand over these systems too. 
 
 2) Hand over the current incomplete (I will document it) conversion scripts that are in my ```charmm36``` branch to @peastman.

**Comment by peastman** [comment]
I still need someone to come up with a set of test cases for patches.  I can verify that the code works as written (I already have tests for that), but I don't know if what I've written is going to meet the requirements.  I don't know enough about either the CHARMM force fields or the intended use cases to know whether it's handling all the situations that are likely to come up.

**Comment by jchodera** [comment]
I believe @ChayaSt said she will generate the test systems for patches for the three use cases that will cover nearly everything people want to do:
* C and N termini
* disulfide bonds
* PMT's (dyes, phosphorylation) I won't have time to write the tests so I will hand over these systems too.

**Comment by jchodera** [comment]
@ChayaSt: Where are you at with things here?

**Comment by rafwiewiora** [comment]
Ok folks, I'm back in the game! I'm going back into fixing impropers assignments, freshening up the amber2openmm script for Amber Tools 16, finishing up the dummy atom metal model support for Zn2+, Ca2+ and Mg2+, in this order. ETA 2 weeks until all functional!

**Comment by jchodera** [comment]
Also congrats to @rafwiewiora on passing his Advancement to Candidacy Exam! 🎉

**Comment by rafwiewiora** [comment]
thanks!

**Comment by peastman** [comment]
Congratulations!

**Comment by swails** [comment]
Congrats!

**Comment by rafwiewiora** [comment]
👍 thanks everyone! :)

**Comment by peastman** [comment]
@rafwiewiora how is this coming?

**Comment by rafwiewiora** [comment]
@peastman : Been back on this full time since yesterday, have the impropers nearly done - writing those suggestions up later today, and then let's see how far I can get by the end of the week!

**Comment by peastman** [comment]
Great!  Thanks for the update!

**Comment by rafwiewiora** [comment]
@peastman @jchodera update from me:

Step 1 out of 3 - new impropers code - done, PR just opened: https://github.com/pandegroup/openmm/pull/1560

Remaining:

Step 2 - try conversion on Amber Tools 16 and modify the conversion script as needed - starting now - no ETA yet as not sure how many problem there will be - let me give an update later today. Meanwhile, looking at the code here: https://github.com/choderalab/openmm/pull/15 would be great if you haven't done so yet. 

Step 3 - write the 'find the amino acids that are ligands for the metal' code, choose the particular models to include, convert to ffxml's.

**Comment by jchodera** [comment]
Thanks for the update, @rafwiewiora!

I think we should discuss the complete plan on Monday, inlcuding:
* code that needs to be written
* how we're going to organize the conversion scripts and input parameter sets
* how we're going to organize the automated testing code

**Comment by rafwiewiora** [comment]
@peastman @jchodera

I'm on vacation and can't join the call today, so update here instead for you: 

1) impropers - TODO: tests: https://github.com/pandegroup/openmm/pull/1560
2) conversion - pretty much done, please review everything, we need to discuss things there, TODO in the PR: https://github.com/pandegroup/openmm/pull/1570
3) metal dummy atom models - all left to do, that is the ligating residue deprotonation code.
